ompi/mca/coll/base/coll_base_alltoall.c

```diff
@@ -297,42 +372,48 @@ int ompi_coll_base_alltoall_intra_bruck(const void *sbuf, int scount,
     }
 
     /* perform communication step */
-    for (distance = 1; distance < size; distance<<=1) {
-
-        sendto = (rank + distance) % size;
-        recvfrom = (rank - distance + size) % size;
+    for (distance = 1; distance < size; distance *= radix) {
+        for (i = 1; i < radix; i++) {
 
-        new_ddt = ompi_datatype_create((1 + size/distance) * (2 + rdtype->super.desc.used));
-
-        /* Create datatype describing data sent/received */
-        for (i = distance; i < size; i += 2*distance) {
-            int nblocks = distance;
-            if (i + distance >= size) {
-                nblocks = size - i;
+            if (distance * i >= size) {
+                break;
             }
-            ompi_datatype_add(new_ddt, rdtype, rcount * nblocks,
-                              i * rcount * rext, rext);
-        }
 
-        /* Commit the new datatype */
-        err = ompi_datatype_commit(&new_ddt);
-        if (err != MPI_SUCCESS) { line = __LINE__; goto err_hndl;  }
+            sendto = (rank + distance * i) % size;
+            recvfrom = (rank - distance * i + size) % size;
 
-        /* Sendreceive */
-        err = ompi_coll_base_sendrecv ( tmpbuf, 1, new_ddt, sendto,
-                                         MCA_COLL_BASE_TAG_ALLTOALL,
-                                         rbuf, 1, new_ddt, recvfrom,
-                                         MCA_COLL_BASE_TAG_ALLTOALL,
-                                         comm, MPI_STATUS_IGNORE, rank );
-        if (err != MPI_SUCCESS) { line = __LINE__; goto err_hndl; }
+            new_ddt = ompi_datatype_create((1 + size/distance) * (2 + rdtype->super.desc.used));
 
-        /* Copy back new data from recvbuf to tmpbuf */
-        err = ompi_datatype_copy_content_same_ddt(new_ddt, 1,tmpbuf, (char *) rbuf);
-        if (err < 0) { line = __LINE__; err = -1; goto err_hndl;  }
+            /* Create datatype describing data sent/received */
+            for (j = i * distance; j < size; j += radix * distance) {
+                int nblocks = distance;
+                if (j + distance >= size) {
+                    nblocks = size - j;
+                }
+                ompi_datatype_add(new_ddt, rdtype, rcount * nblocks,
+                                  j * rcount * rext, rext);
+            }
 
-        /* free ddt */
-        err = ompi_datatype_destroy(&new_ddt);
-        if (err != MPI_SUCCESS) { line = __LINE__; goto err_hndl;  }
+            /* Commit the new datatype */
+            err = ompi_datatype_commit(&new_ddt);
+            if (err != MPI_SUCCESS) { line = __LINE__; goto err_hndl;  }
+
+            /* Sendreceive */
+            err = ompi_coll_base_sendrecv ( tmpbuf, 1, new_ddt, sendto,
```

Comment:
these can be replaced with non-blocking send/recv for better performance

Reply:
I tried replacing with non-blocking send/recv but it segfaults in osu_alltoall

Reply:
You won't get any performance buff with large fanout if you are using blocking send/receives.  We can setup a meeting next week to see why it crashes.

Reply:
You do get some improvement because you decrease the depth of the tree, but you don't get the multi-rail benefit.

Talking about multi-rail, I would expect the p2p communication layer to take advantage of the multi-rail, at least after a certain message size. If that's the case when the upper and lower level of the communication stack both try to generate multi-rail traffic things cant go well. How do we protect against that ? Is the plan to only enable collectives multi-rail only for some ranges of messages ?

Reply:
Jiaxiyan is on an international trip.  I will follow up with her to resolve the crash she mentioned.